configfile: "/scratch1/fs1/martyomov/carisa/snakemake-tutorial/config/config.yaml"

rule all:
    input:
        "results/plots/quals.svg"

rule bwa_map:
    input:
        "resources/data/genome.fa",
        lambda wildcards: config["samples"][wildcards.sample]
    output:
        "results/mapped_reads/{sample}.bam"
    conda: "./envs/bwa.yaml"
    shell:
        "bwa mem {input} | samtools view -Sb - > {output}"
rule samtools_sort:
    input:
        "results/mapped_reads/{sample}.bam"
    output:
        "results/sorted_reads/{sample}.sorted.bam" #made my own edit to the names
    conda: "./envs/bwa.yaml"
    shell:
        """
        samtools sort -T sorted_reads/{wildcards.sample} -O bam {input} > {output}
        """

rule samtools_index:
    input:
        "results/sorted_reads/{sample}.sorted.bam"
    output:
        "results/sorted_reads/{sample}.sorted.bam.bai"
    conda: "./envs/bwa.yaml"
    shell:
        "samtools index {input}"

rule bcf_call:
    input:
        fa="resources/data/genome.fa",
        bam=expand("results/sorted_reads/{sample}.sorted.bam",sample=config["samples"]),
        bai=expand("results/sorted_reads/{sample}.sorted.bam.bai", sample=config["samples"])
    output:
        "results/calls/all.vcf"
    conda: "./envs/bcf.yaml"
    shell:
        """
        bcftools mpileup -f {input.fa} {input.bam} | bcftools call -mv - > {output}
        """

rule plot_quals:
    input:
        "results/calls/all.vcf"
    output:
        "results/plots/quals.svg"
    conda: "./envs/environment.yaml"
    script:
        "scripts/plot-quals.py"

